prov-api/helpers/prov_connector.py

```diff
@@ -0,0 +1,1597 @@
+
+    # The query to see effected agents one step removed from any downstream occurrences i.e. 'effected people/organisations'
+    query = f"""
+    MATCH r=((agent: Agent) <-[]- (downstream) -[*0..{depth}]->(start{{`{IDENTIFIER_TAG}`:'{starting_id}'}})) RETURN r
+    """
+
+    # Run the query
+    try:
+        result = run_query(query, config)
+    except Exception as e:
+        raise HTTPException(
+            status_code=500,
+            detail=f"Neo4j lineage query experienced an exception: {e}."
+        )
+
+    # Convert to list of paths by pulling out the 'r' field from the RETURN r in above
+    # cypher query
+    paths: List[Optional[graph.Path]] = [record.get('r') for record in result]
+
+    # Apply simple exploration algorithm to traverse paths and produce networkx digraph
+    graph = generate_networkx_graph_from_lineage_paths(
+        lineage_paths=paths
+    )
+
+    # Return D3.js friendly serialisation
+    return networkx.node_link_data(graph)
+
+
+def upstream_query(starting_id: str, depth: int, config: Config) -> Dict[str, Any]:
+    """    lineage_query
+        Performs a depth limited upstream query for a neo4j database. 
+        This query is defined as, starting at a node identified uniquely by
+        a handle id, traverse any relation in the parent direction i.e. 
+        (parent) <-[]-(starting_node).
+
+        The multiplicity of this exploration step is bounded by the depth 
+        provided. 
+
+        The result from the neo4j query is a list of records. A record is a kv
+        store of returned elements from the cypher query. This is parsed into a 
+        list of path objects. A path is a start, end and a list of steps/relationships
+        along the way. 
+
+        These paths are explored to generate a graph. The graph is a Networkx.DiGraph. 
+        This digraph can be returned in a D3.js friendly json serialisation format.        
+
+        Arguments
+        ----------
+        starting_id : str
+            The handle ID which should uniquely identify a node in the graph
+        depth : int
+            The maximum depth to traverse in the <-[*1..depth]- wildcard 
+        config : Config
+            Connection config etc
+
+        Returns
+        -------
+         : Dict[str, Any]
+            The DiGraph serialisation format which is D3.js friendly.
+            This format is JSON and contains a list of nodes with an id, 
+            category and subtype + a list of links which have a from, 
+            to and a relationship type.
+
+        See Also (optional)
+        --------
+
+        Examples (optional)
+        --------
+    """
+    if config.mock_graph_db:
+        raise RuntimeError(
+            "Asking for real data from graph DB during mock! Returning [].")
+
+    # The cypher query to make for upstream lineage
+    query = f"""
+    MATCH r=((parent) <-[*1..{depth}]-(start{{`{IDENTIFIER_TAG}`:'{starting_id}'}})) RETURN r
+    """
```

Comment:
what if all these queries were defined somewhere else or generated from functions that take in starting id, parents etc. Currently the query variable is the only thing different between all these *_query functions. So it would be easy to make a general "execute query" function that maybe recieves a query and then performs the mock db check, executes query, gets paths, gets graph, and returns. Maybe im missing something?

Reply:
i guess it can be nice to have a specific function name for a query type so i understand that use case for having a function for each query

Reply:
This is all old code - out of scope for now. But noted.